test/GroupByTest.cpp

```diff
@@ -421,7 +445,100 @@ TEST_F(GroupBySpecialCount, getPermutationForThreeVariableTriple) {
+              avgAggregate);
+  // Must be AVG aggregate (for now)
+  testFailure(variablesOnlyX, aliasesCountX, validJoinWhenGroupingByXPlusSort,
+              countAggregate);
+  // Top operation must be SORT
+  testFailure(variablesOnlyX, aliasesAvgX, validJoinWhenGroupingByX,
+              avgAggregate);
+  // Can not be a nested aggregate
+  testFailure(variablesOnlyX, aliasesAvgCountX,
+              validJoinWhenGroupingByXPlusSort, avgCountAggregate);
+  // Optimization has to be enabled
+  RuntimeParameters().set<"use-group-by-hash-map-optimization">(false);
+  testFailure(variablesOnlyX, aliasesAvgX, validJoinWhenGroupingByXPlusSort,
+              avgAggregate);
+
+  // Everything is valid for the following example.
+  RuntimeParameters().set<"use-group-by-hash-map-optimization">(true);
+  GroupBy groupBy{qec, variablesOnlyX, aliasesAvgX,
+                  validJoinWhenGroupingByXPlusSort};
+  auto optimizedAggregateData =
+      groupBy.checkIfHashMapOptimizationPossible(avgAggregate);
+  ASSERT_TRUE(optimizedAggregateData.has_value());
+  ASSERT_EQ(optimizedAggregateData->subtreeColumnIndex_, 0);
+}
+
+// _____________________________________________________________________________
+TEST_F(GroupByOptimizations, correctResultForHashMapOptimization) {
+  /* Setup query:
+  SELECT ?x (AVG(?y) as ?avg) WHERE {
+    ?z <is-a> ?x .
+    ?z <is> ?y
+  } GROUP BY ?x
+ */
+  Tree zxScan = makeExecutionTree<IndexScan>(
+      qec, Permutation::Enum::PSO,
+      SparqlTriple{Variable{"?z"}, {"<is-a>"}, Variable{"?x"}});
+  Tree zyScan = makeExecutionTree<IndexScan>(
+      qec, Permutation::Enum::PSO,
+      SparqlTriple{Variable{"?z"}, {"<is>"}, Variable{"?y"}});
+  Tree join = makeExecutionTree<Join>(qec, zxScan, zyScan, 0, 0);
+  std::vector<ColumnIndex> sortedColumnsVector = {1};  // is ?x in column 1?
+  Tree sortedJoin = makeExecutionTree<Sort>(qec, join, sortedColumnsVector);
+
+  SparqlExpressionPimpl avgYPimpl = makeAvgPimpl(varY);
+  std::vector<Alias> aliasesAvgY{Alias{avgYPimpl, Variable{"?avg"}}};
+  std::vector<GroupBy::Aggregate> avgAggregate = {{avgYPimpl, 1}};
+
+  // Calculate result with optimization
+  RuntimeParameters().set<"use-group-by-hash-map-optimization">(true);
+  GroupBy groupByWithOptimization{qec, variablesOnlyX, aliasesAvgY, sortedJoin};
+  auto resultWithOptimization = groupByWithOptimization.getResult();
+
+  // Clear cache, calculate result without optimization
+  qec->clearCacheUnpinnedOnly();
+  RuntimeParameters().set<"use-group-by-hash-map-optimization">(false);
+  GroupBy groupByWithoutOptimization{qec, variablesOnlyX, aliasesAvgY,
+                                     sortedJoin};
+  auto resultWithoutOptimization = groupByWithoutOptimization.getResult();
+
+  // Compare results, using debugString as the result only contains 2 rows
+  ASSERT_EQ(resultWithOptimization->asDebugString(),
+            resultWithoutOptimization->asDebugString());
+}
```

Comment:
You shouldn't use debug strings for such important things as testing.

Best you write a gmock matcher (have a look at the `gmock/gtest cookbock on the internet, 
or for example look at the `QueryPlannerTestHelpers.h` file for some examples.
These have the advantage (over implementing a manual equality check) that you get better error messages on failure.